### PR TITLE
Fix build with sphinx 7

### PR DESCRIPTION
**Pull request by antonio-rojas**
Replace or remove use of deprecated stuff

Tried to keep it compatible with older sphinx, but needs testing

### CODE REVIEW

**Comment by kiwifb** [comment]
Thanks for this. I looked at things when sphinx 6 came out and just gave up. `sage_autodoc.py` needs to die one day soon.

File: src/sage_docbuild/ext/sage_autodoc.py [hunk 1]
```diff
@@ -56,8 +55,11 @@
...
                                  stringify_signature)
 from sphinx.util.typing import OptionSpec, get_type_hints, restify
-from sphinx.util.typing import stringify as stringify_typehint
-
+try:
+    from sphinx.util.typing import stringify_annotation
+except ImportError:
+    from sphinx.util.typing import stringify as stringify_annotation
+    
 # ------------------------------------------------------------------
```
Comment (kwankyu) [inline src/sage_docbuild/ext/sage_autodoc.py:65]: There are invisible white spaces in L62. Remove them.

I would put a blank line before `try:`

File: src/sage_docbuild/ext/sage_autodoc.py [hunk 2]
```diff
@@ -645,34 +647,6 @@ def add_content(self, more_content: Optional[StringList]) -> None:
...
-                    logger.warning(__('missing attribute %s in object %s') %
-                                   (name, self.fullname), type='autodoc')
-            return False, selected
-        elif self.options.inherited_members:
-            return False, [(m.name, m.value) for m in members.values()]
-        else:
-            return False, [(m.name, m.value) for m in members.values()
-                           if m.directly_defined]
-
     def filter_members(self, members: ObjectMembers, want_all: bool
```
Comment (kwankyu) [inline src/sage_docbuild/ext/sage_autodoc.py:661]: They did not wipe this out completely but left
```py
    def get_object_members(self, want_all: bool) -> tuple[bool, ObjectMembers]:
        """Return `(members_check_module, members)` where `members` is a
        list of `(membername, member)` pairs of the members of *self.object*.

        If *want_all* is True, return all members.  Else, only return those
        members given by *self.options.members* (which may also be None).
        """
        raise NotImplementedError('must be implemented in subclasses')                                                                   
```
If we want `sage_autodoc.py` to be as close to the original as possible, we may put this definition too.

File: src/sage_docbuild/conf.py [hunk 3]
```diff
@@ -248,7 +248,7 @@ def set_intersphinx_mappings(app, config):
         if os.path.isdir(os.path.join(invpath, directory)):
             src = os.path.join(refpath, directory)
             dst = os.path.join(invpath, directory, 'objects.inv')
-            app.config.intersphinx_mapping[src] = dst
+            app.config.intersphinx_mapping[directory] = (src, dst)
```
Comment (kwankyu) [inline src/sage_docbuild/conf.py:250]: `directory` is used as a name for `intersphinx_mapping`? Then perhaps you may want to edit the comment starting from L240.

**Comment by kwankyu** [comment]
The built documents look good to me.

File: src/sage_docbuild/conf.py (hunk 3 above)
Comment (antonio-rojas) [reply src/sage_docbuild/conf.py:250]: It is supposed to be a unique identifier, this seemed like the most natural choice.

**Comment by kwankyu** [comment]
Thanks. But now we have
```
[sagemath_doc_html-none] [reference]   File "/__w/sage/sage/src/sage_docbuild/ext/sage_autodoc.py", line 651, in Documenter
[sagemath_doc_html-none] [reference]     def get_object_members(self, want_all: bool) -> tuple[bool, ObjectMembers]:
[sagemath_doc_html-none] [reference] TypeError: 'type' object is not subscriptable
```

**Comment by kwankyu** [comment]
It seems "tuple" type annotation is only introduced in python 3.9, and "Tuple" is deprecated since python 3.9.

**Comment by mkoeppe** [comment]
Per https://doc.sagemath.org/html/en/developer/coding_in_python.html#python-language-standard, this file should use `from __future__ import annotations`

**Review by kwankyu — APPROVED** [review]
LGTM.

**Comment by kiwifb** [comment]
I have successfully build the doc in sage-on-gentoo with both sphinx 5 and sphinx 7. Definitely happy with the PR.

**Comment by tobiasdiez** [comment]
Are all build issues with sphinx 7 fixed by this PR? If yes, can we remove the version constraints in https://github.com/sagemath/sage/blob/develop/build/pkgs/sphinx/install-requires.txt and https://github.com/sagemath/sage/blob/develop/build/pkgs/sphinx/distros/conda.txt?

**Comment by kiwifb** [comment]
> Are all build issues with sphinx 7 fixed by this PR? If yes, can we remove the version constraints in https://github.com/sagemath/sage/blob/develop/build/pkgs/sphinx/install-requires.txt and https://github.com/sagemath/sage/blob/develop/build/pkgs/sphinx/distros/conda.txt?

I haven't encountered any issues with sphinx 7. So, I would say you should go ahead with that, but in a separate PR since this one is being merged.

**Comment by dimpase** [comment]
We definitely should update our sphinx, as it blocks conda (they need >= 6.0 for some Sage deps) @isuruf

**Comment by tobiasdiez** [comment]
> > Are all build issues with sphinx 7 fixed by this PR? If yes, can we remove the version constraints in https://github.com/sagemath/sage/blob/develop/build/pkgs/sphinx/install-requires.txt and https://github.com/sagemath/sage/blob/develop/build/pkgs/sphinx/distros/conda.txt?
> 
> I haven't encountered any issues with sphinx 7. So, I would say you should go ahead with that, but in a separate PR since this one is being merged.

Done now at https://github.com/sagemath/sage/pull/35845 (ready for review).

**Comment by dimpase** [comment]
I seem to have a problem with sphinx 7.2.6:
```
[sagemath_doc_html-none] [plotting ] /home/dimpase/work/software/sage/src/sage/plot/contour_plot.py:docstring of sage.plot.contour_plot.region_plot:1: WARNING: could not parse arglist ("f, xrange, yrange, plot_points, incol, outcol, bordercol, borderstyle, borderwidth, alpha, plot_points=100, incol='blue', outcol=None, bordercol=None, borderstyle=None, borderwidth=None, frame=False, axes=True, legend_label=None, aspect_ratio=1, alpha=1, **options"): duplicate parameter name: 'plot_points'
```

has anyone seen this? Looks like it complains about `@options(plot_points=100,...`
decorator on `region_plot`.

**Comment by antonio-rojas** [comment]
> has anyone seen this? Looks like it complains about `@options(plot_points=100,...` decorator on `region_plot`.

Yes. For now I'm whitelisting that warning in `sage_docbuild/sphinxbuild.py`, since I don't really understand decorators.

**Comment by kiwifb** [comment]
I have seen trouble with both 7.1 and 7.2. In sage-on-gentoo anything sphinx 7.1 and over breaks the build last time I checked. See https://github.com/cschwan/sage-on-gentoo/issues/753

**Comment by dimpase** [comment]
I can confirm that sphinx 7.0.0 (with docutils 0.19) still works for me (in a pyenv and Python 3.11.5 there, in conjunction with trying out #36256), but 7.1.0 (with docutils 0.20) does not.

**Comment by antonio-rojas** [comment]
> I have seen trouble with both 7.1 and 7.2. In sage-on-gentoo anything sphinx 7.1 and over breaks the build last time I checked. See [cschwan/sage-on-gentoo#753](https://github.com/cschwan/sage-on-gentoo/issues/753)

I can't reproduce this with 7.2.6

**Comment by dimpase** [comment]
I'll try from scratch, again.

**Comment by dimpase** [comment]
but what I saw was a different error

**Comment by kiwifb** [comment]
Could be a combination of sphinx with something else that fails. Right now furo-2023-8.19 leads to truncated html documentation for me as well. The documentation build is just in a bad place right now.

**Comment by kwankyu** [comment]
Please open a new issue.

**Pull request merged**